### DISCUSSION TITLE
Setup IKBT Envirment in my Win10 computer with conda and pycharm

### DISCUSSION THREAD

**Discussion by 1127378627** [uw-biorobotics/IKBT#43] [General]
The python version is  2.7. When I run ikSolver.py, some errors I cannot solve:
C:\Users\hanyj\Anaconda3\envs\IKBT\python.exe "C:/Users/hanyj/OneDrive - mails.jlu.edu.cn(1)/HCADCAM/Poli_solution/ikbt/ikSolver.py"
Traceback (most recent call last):
  File "XX/HCADCAM/Poli_solution/ikbt/ikSolver.py", line 30, in <module>
    import ikbtfunctions.output_python as op
  File "XX\HCADCAM\Poli_solution\ikbt\ikbtfunctions\output_python.py", line 51
    ''', file=f)
             ^
SyntaxError: invalid syntax

Process finished with exit code 1

Can anyone help to test the project in pycharm?

**Comment by blake5634** [comment]
Python 3 is required.   You can find some earlier versions where the code worked with Python 2.7

**Reply by blake5634** [reply]
Han,
  I'm sorry I don't have time to debug this deeply, but to test your solution I suggest starting with a set of joint angles (a pose), and then running forward kinematics (generated by ikbt) to generate Td.   Then run inverse kinematics on Td and see if solutions return your original pose.  That way you are guaranteed that Td is a reachable pose. 
BH

**Comment by 1127378627** [comment]
Dear Professor,
Thank you very much! You are right. I find that the forward solution is different from what I expected. This may be the problem!
I will check this deeply. Are there any special requirements for DH establishment rules of this algorithm?

**Reply by blake5634** [reply]
No special requirements.    Please "git pull" the code because I corrected a bug in the Latex output yesterday.

**Reply by 1127378627** [reply]
Dear Professor,
I solved my problem. In this code,  the MDH table should be used, not SDH. Thank you for your help!
Best regards,
Han

**Comment by blake5634** [comment]
Great,
Did you get a useful IK solution?
  Would you mind sending me the final working robot setup for your Kuka?
 (i.e. your new entry in ik_robots.py)
BH

On Fri, Nov 12, 2021 at 7:24 AM 1127378627 ***@***.***> wrote:

> Dear Professor,
> I solved my quesstion. In this code, the MDH table should be used, not
> SDH. Thank you for your help!
> Best regards,
> Han
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/uw-biorobotics/IKBT/discussions/43#discussioncomment-1632525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRXOSL2IQAL4EWVZNJGQQ3ULUWTXANCNFSM5HH4NRGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Blake Hannaford
--

**Reply by 1127378627** [reply]
Dear professor，
I have not yet solved the IK solution, just the forward problem.
Now I am trying to solve IK solution, but in calculation, many tips like this:
```
Assigner: variable on blackboard: th_3
algebra_id: there are no one-unknown equations

```
Han

**Reply by 1127378627** [reply]
Dear professor，
The follow configuration  for Kuka KR16 can successfully solve the IK solution 

```
        dh = sp.Matrix([
            [       0,     0,     h,             th_1],
            [-sp.pi/2,   a_1,     0,             th_2+sp.pi/2],
            [       0,   a_2,     0,             th_3],
            [sp.pi/2,    a_3,   d_4,             th_4],
            [-sp.pi/2,     0,     0,             th_5],
            [sp.pi/2,      0,     0,             th_6],
        ])

        sp.var('h a_1')
        vv = [1, 1, 1, 1, 1, 1]
        variables = [unknown(th_1), unknown(th_2), unknown(th_3), unknown(th_4), unknown(th_5), unknown(th_6)]

        params = [a_1, h, a_2, a_3, d_4, d_6]
        pvals = {a_1: 0.26, h: 0.675, a_2: -0.68, a_3: 0.035, d_4: 0.67, d_6: 0.158}
```
For home position, it will cause error to solve the IK, maybe some bugs. And sometimes nan comes in the solution. I replace them with zero.
Han YJ
